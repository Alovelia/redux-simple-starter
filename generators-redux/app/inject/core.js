let path = require('path');
let chalk = require('chalk');
let fse = require('fs-extra');
let fs = require('fs');
let Promise = require('bluebird');
let config = require('../config');
let _ = require('../utils/mixins');
let rules = require('./rules');

let log = console.log.bind(console);
let readFile = Promise.promisify(fs.readFile);
let writeFile = Promise.promisify(fs.writeFile);

async function injectCode(props) {
  let target = path.resolve(
    config.APP_ROOT,
    props.wrapperOuter || '',
    props.moduleName || '',
    props.wrapperInner || '',
    (props.fileName || '') + (props.ext || '.js')
  );

  console.log(chalk.yellow('Injecting to: ') + chalk.gray(target));

  try {
    if (!await fse.pathExists(target)) {
      throw new Error(`${target} doesn't exist. Create it before injecting`);
    }
    const data = await readFile(target);
    const newData = await makeInjection(data, props);
    await writeFile(target, newData);
  } catch (e) {
    console.info(chalk.red(e.message));
  }
  return target;
}

function makeInjection(content, props) {
  let settings = Object.assign({}, _, props);
  /* Any modification goes here. Note that contents is a Buffer object */
  let newContent = content.toString();
  let isMultiple = !!~props.name.indexOf(' ');
  let injections = [props.name];

  if (isMultiple) {
    injections = props.name.split(' ');
    // if some keys are also arrays and should be used during injection in the same order as name: '1 2 3', key2:
    // 'v1 v2 v3'
    if (settings.multiKeys) {
      settings.multiKeys.forEach((key) => {
        settings[key] = settings[key].split(' ');
      });
    }
  }
  injections.forEach((injection, index) => {
    props.rules.forEach((rule) => {
      let match = rule.cursor;
      let multipleObj = Object.assign({}, settings);
      if (isMultiple && settings.multiKeys) {
        settings.multiKeys.forEach((key) => {
          multipleObj[key] = settings[key][index];
        });
      }
      let replacement = _.template(rule.replacement)(
        Object.assign(multipleObj, { name: injection })
      );
      let regEx = new RegExp(`(\\s*|\\t*)(${match})`, 'g');
      newContent = newContent.replace(regEx, `$1${replacement}$1$2`);
    });
  });
  return newContent;
}

async function injector(props, options) {
  let settings = Object.assign({}, _, props, options, { rules: rules[options.rules] });
  return injectCode(settings);
}

exports.injector = _.curry(injector);
