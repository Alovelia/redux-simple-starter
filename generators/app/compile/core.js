let chalk = require('chalk'),
  { copyTpl } = require('../utils/copy'),
  _ = require('../utils/mixins'),
  config = require('../config'),
  path = require('path');

//It compiles file from corresponding template to corresponding destination
async function compileTemplate(props) {
  let destination = path.resolve(
      config.APP_ROOT,
      props.wrapperOuter || '',
      props.moduleName || '',
      props.wrapperInner || ''
    ),
    fileName = path.resolve(destination, (props.fileName || props.name) + (props.ext || '.js')),
    templateName = path.resolve(config.TEMPLATE_ROOT, props.tmplName);

  await copyTpl(templateName, fileName, props);
  return fileName;
}

async function compiler(props, options) {
  let settings = Object.assign({}, props, options);

  if (!~settings.name.indexOf(' ')) {
    // create single instance
    return compileTemplate(settings);
  }
  // create multiple instances
  return settings.name.split(' ')
    .map(async (name, index) => {
      let multiObj = {};
      // It relies on `multiOption` method for keys which were defined in `multiKeys` option
      if (options.multiKeys) {
        // each element inside this `multiple` array should be set via `multiOption` method
        options.multiKeys.forEach((key) => {
          multiObj[key] = options[key][index];
        });
      }
      // this way source option with multiple values will be replaced with single value for current index
      let settingsSingle = Object.assign({}, props, options, { name }, multiObj);
      return compileTemplate(settingsSingle);
    });
}

// for compilers which could support multiple strategies of creation using 'space' as a separator
// For example user inputs 'component1 component2' as a props.name in terminal
//   folderName: multiOption(props.name, 'components/'), // if no space then it will be just folderName: 'components/component1'
//   folderName: multiOption(props.name, 'components/'), // if space it will be folderName: ['components/component1', 'components/component2']
//   fileName: multiOption(props.name, null, '.mod'),
//   multiple: ['folderName', 'fileName']
// It will create
//   folderName: `${props[]}`
function multiOption(value, prefix, sufix) {
  return ~value.indexOf(' ')
    ? value
      .split(' ')
      .map(single => (`${prefix || ''}${single}${sufix || ''}`))
    : `${prefix || ''}${value}${sufix || ''}`;
}

exports.compiler = _.curry(compiler);
exports.multiOption = multiOption;
