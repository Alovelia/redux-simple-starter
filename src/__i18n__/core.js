import { addLocaleData } from 'react-intl';
import isIntlLocaleSupported from 'intl-locales-supported';

/*
* This object contains async functions
* where keys correspond to user language
* Each function lazy loads scripts related to i18n
* and messages for particular language.
* */
export const localeLoaders = {
  en: async () => {
    if (global.Intl && isIntlLocaleSupported('en')) {
      const [locale, messages] = await Promise.all([
        import('react-intl/locale-data/en' /* webpackChunkName: "intl-locale-en" */),
        import('./en.json' /* webpackChunkName: "locale-en" */)
      ]);
      addLocaleData(locale);
      return messages;
    }
    // polyfill for old browsers
    const [, locale, messages] = await Promise.all([
        import('intl' /* webpackChunkName: "intl-polyfill" */),
        import('react-intl/locale-data/en' /* webpackChunkName: "intl-locale-en" */),
        import('./en.json' /* webpackChunkName: "locale-en" */)
    ]);
    await import('intl/locale-data/jsonp/en-GB' /* webpackChunkName: "intl-polyfill" */);
    addLocaleData(locale);
    return messages;
  },
  // to add another language - uncomment the next block
//   de: async () => {
//     if (global.Intl && isIntlLocaleSupported('de')) {
//       const [locale, messages] = await Promise.all([
//         import('react-intl/locale-data/de' /* webpackChunkName: "intl-locale-de" */),
//         import('./de.json' /* webpackChunkName: "locale-de" */),
//         import('moment/locale/de' /* webpackChunkName: "intl-locale-de" */)
//       ]);
//       addLocaleData(locale);
//       return messages;
//     }
//     // polyfill for old browsers
//     const [, locale, messages] = await Promise.all([
//         import('intl' /* webpackChunkName: "intl-polyfill" */),
//         import('react-intl/locale-data/de' /* webpackChunkName: "intl-locale-en" */),
//         import('./de.json' /* webpackChunkName: "locale-de" */),
//         import('moment/locale/de' /* webpackChunkName: "intl-locale-en" */)
//     ]);
//     await import('intl/locale-data/jsonp/de' /* webpackChunkName: "intl-polyfill" */);
//     addLocaleData(locale);
//     return messages;
//   },
};

/*
* It loads all locale data dependencies and messages
* for chosen locale using lazy load
* @returns {Promise} - resolved with messages
* */
export function loadLocaleData(locale) {
  return localeLoaders[locale]();
}

/*
* @param {string} locale - Current locale e.g 'EN'
* @returns {Bool}
* */
export function isLocaleSupported(locale) {
  return !!localeLoaders[locale];
}

/*
* Define user's language. Different browsers have the user locale defined
* on different fields on the `navigator` object, so we make sure to account
* for these different by checking all of them
* @returns {string} 'en-US'
* */
export function getDefaultLanguage() {
  let navigator = global.navigator;
  return (navigator.languages && navigator.languages[0])
    || navigator.language
    || navigator.userLanguage;
}

/*
* @param {string} language - Current language e.g 'en-US'
* @returns {string} 'en' - trimmed navigator language
* */
export function getLangWithoutRegionCode(language) {
  return language.toLowerCase().split(/[_-]+/)[0];
}

/*
* @returns {string} userLang or default - language supported by the app
* */
export function getDefaultSupportedLanguage() {
  const DEFAULT_LANG = 'en';
  const userLang = getLangWithoutRegionCode(getDefaultLanguage());

  if (isLocaleSupported(userLang)) {
    return userLang;
  }

  return DEFAULT_LANG;
}
