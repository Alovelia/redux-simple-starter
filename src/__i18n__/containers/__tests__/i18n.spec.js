import React from 'react';
import { shallowWithIntl } from 'enzyme-react-intl';
import { shallow, mount } from 'enzyme';
import { createMockStore } from 'redux-test-utils';
import { fromJS } from 'immutable';
import sinon from 'sinon';
import { FormattedMessage, defineMessages } from 'react-intl';
import I18nConected, { I18nContainer } from '../i18n';
import * as localeCore from '../../core';

describe('<I18nContainer />', () => {
  context('when no messages', () => {
    let wrapper;
    const updateMessages = sinon.spy();
    let loadLocaleDataMock;
    beforeAll(() => {
      const children = <h1>TEST</h1>;
      const props = {
        messages: {},
        locale: 'en',
        updateMessages
      };
      loadLocaleDataMock = sinon
        .stub(localeCore, 'loadLocaleData')
        .returns(Promise.resolve(props.messages));
      wrapper = shallow(<I18nContainer {...props}>{children}</I18nContainer>);
    });

    afterAll(() => {
      loadLocaleDataMock.restore();
    });

    it('should be empty if no messages', () => {
      expect(wrapper.text()).to.equal('');
    });
  });

  context('when just mounted', () => {
    let wrapper;
    let loadLocaleDataMock;
    // const updateMessages = jest.fn();
    const updateMessages = sinon.spy();
    const messages = { id: 'test' };
    const children = <h1>TEST</h1>;
    const props = {
      messages: {},
      locale: 'en',
      updateMessages
    };
    beforeAll(() => {
      loadLocaleDataMock = sinon
        .stub(localeCore, 'loadLocaleData')
        .returns(Promise.resolve(messages));
      shallow(<I18nContainer {...props}>{children}</I18nContainer>);
    });

    afterAll(() => {
      loadLocaleDataMock.restore();
    });
    it('should call updateMessages action in componentDidMount', () => {
      // expect(updateMessages).toHaveBeenCalledWith(messages);
      expect(loadLocaleDataMock).to.have.been.called();
      expect(loadLocaleDataMock).to.have.been.calledWith(props.locale);
      expect(updateMessages).to.have.been.called();
      expect(updateMessages).to.have.been.calledWith(messages);
    });
  });

  context('should render translated message', () => {
    let loadLocaleDataMock;
    // const updateMessages = jest.fn();
    const updateMessages = sinon.spy();
    const messages = { id: 'test' };
    let wrapper;

    const props = {
      messages,
      locale: 'en',
      updateMessages
    };
    beforeAll(() => {
      loadLocaleDataMock = sinon
        .stub(localeCore, 'loadLocaleData')
        .returns(Promise.resolve(messages));
      wrapper = shallow(<I18nContainer {...props}><FormattedMessage id="id" /></I18nContainer>);
    });

    afterAll(() => {
      loadLocaleDataMock.restore();
    });
    it('should render with proper translation', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render proper message', () => {
      expect(wrapper.html()).to.equal(`<span>${props.messages.id}</span>`);
    });
  });

  context('should switch language', () => {
    let loadLocaleDataMock;
    const updateMessages = sinon.spy();
    const messages = { id: 'test' };
    let wrapper;

    const props = {
      messages,
      locale: 'en',
      updateMessages
    };
    const newProps = {
      messages: { id: 'тест' },
      locale: 'ru',
      updateMessages
    };
    beforeAll(() => {
      loadLocaleDataMock = sinon
        .stub(localeCore, 'loadLocaleData')
        .onCall(0)
        .returns(Promise.resolve(messages))
        .onCall(1)
        .returns(Promise.resolve(newProps.messages));
    });

    afterAll(() => {
      loadLocaleDataMock.restore();
    });
    it('should render proper message', async () => {
      sinon.spy(I18nContainer.prototype, 'componentDidMount');
      sinon.spy(I18nContainer.prototype, 'componentWillReceiveProps');
      wrapper = shallow(<I18nContainer {...props}><FormattedMessage id="id" /></I18nContainer>);
      expect(I18nContainer.prototype.componentDidMount.calledOnce).to.equal(true);
      await wrapper.setProps(newProps);
      // wrapper.update();
      expect(I18nContainer.prototype.componentWillReceiveProps.calledOnce).to.equal(true);
      expect(loadLocaleDataMock).to.have.been.called();
      // expect(loadLocaleDataMock).to.have.been.calledTwice();
      expect(loadLocaleDataMock).to.have.been.callCount(2);
      expect(loadLocaleDataMock).to.have.been.calledWith(newProps.locale);
      expect(updateMessages).to.have.been.called();
      expect(updateMessages).to.have.been.calledWith(messages);
      expect(wrapper.html()).to.equal(`<span>${newProps.messages.id}</span>`);
    });
  });
});
