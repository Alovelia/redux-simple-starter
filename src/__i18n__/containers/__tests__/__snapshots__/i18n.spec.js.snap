// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<I18nContainer /> should render translated message should render with proper translation 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <I18nContainer
    locale="en"
    messages={
      Object {
        "id": "test",
      }
    }
    updateMessages={[Function]}
  >
    <FormattedMessage
      id="id"
      values={Object {}}
    />
  </I18nContainer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "en",
    "nodeType": "class",
    "props": Object {
      "children": <FormattedMessage
        id="id"
        values={Object {}}
      />,
      "locale": "en",
      "messages": Object {
        "id": "test",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "id": "id",
        "values": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "en",
      "nodeType": "class",
      "props": Object {
        "children": <FormattedMessage
          id="id"
          values={Object {}}
        />,
        "locale": "en",
        "messages": Object {
          "id": "test",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "id": "id",
          "values": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
